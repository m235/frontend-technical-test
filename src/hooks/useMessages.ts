import type { AxiosError } from 'axios'
import { useQuery } from 'react-query'

import { getMessages, MessagesApi } from '@/services/messages'
import type { Message } from '@/types/message'

const useMessages = (conversationId: number) => {
  return useQuery<Message[], AxiosError>([MessagesApi.MESSAGES, conversationId], () => getMessages(conversationId), {
    enabled: !!conversationId,
    useErrorBoundary: true,
  })
}

export default useMessages
