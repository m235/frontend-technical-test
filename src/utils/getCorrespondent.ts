import type { Conversation } from '@/types/conversation'

export const getCorrespondent = (userId: number, conversation: Conversation) => {
  switch (userId) {
    case conversation.senderId:
      return conversation.recipientNickname
    case conversation.recipientId:
      return conversation.senderNickname
    default:
      throw new Error('You are nor the recipient nor the sender')
  }
}
