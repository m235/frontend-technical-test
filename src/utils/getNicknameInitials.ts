export const getNicknameInitials = (nickname: string) =>
  nickname
    .split(' ')
    .map((string) => string.trim().charAt(0))
    .slice(0, 2)
    .join('')
