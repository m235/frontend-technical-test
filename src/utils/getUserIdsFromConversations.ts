import type { Conversation } from '@/types/conversation'

export const getUserIdsFromConversations = (conversations?: Conversation[]) =>
  Array.from<number>(new Set(conversations?.reduce((acc, curr) => [...acc, curr.recipientId, curr.senderId], []) ?? []))
