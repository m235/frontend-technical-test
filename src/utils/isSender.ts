import type { Conversation } from '@/types/conversation'

export const isSender = (userId: number, conversation: Conversation) => userId === conversation.senderId
