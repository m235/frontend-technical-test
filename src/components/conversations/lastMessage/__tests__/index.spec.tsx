import LastMessage from '@/components/conversations/lastMessage'
import useMessages from '@/hooks/useMessages'
import { render } from '@/tests/utils'
import type { Conversation } from '@/types/conversation'
import type { Message } from '@/types/message'

const mockedUseMessages = useMessages as jest.Mock<any>
jest.mock('@/hooks/useMessages')

describe('components:conversations:lastMessage', () => {
  const conversation: Conversation = { id: 1, recipientNickname: 'John', recipientId: 1, senderNickname: 'Travolta', senderId: 2 }
  const messages: Message[] = [
    { id: 1, timestamp: new Date().getTime(), authorId: 1, conversationId: 1, body: 'Foo' },
    { id: 2, timestamp: new Date().getTime(), authorId: 2, conversationId: 1, body: 'Bar' },
  ]

  beforeEach(() => {
    mockedUseMessages.mockImplementation(() => ({ isLoading: true }))
  })
  afterEach(() => {
    jest.clearAllMocks()
  })
  it('should render default correctly', async () => {
    const { container } = render(<LastMessage conversationId={conversation.id} />)

    expect(container).toMatchSnapshot()
  })

  it('should render the loading component', async () => {
    const { getByTestId } = render(<LastMessage conversationId={conversation.id} />)

    expect(getByTestId('loading')).toBeTruthy()
  })

  it('should render the last message body and date', async () => {
    mockedUseMessages.mockImplementation(() => ({ isLoading: false, data: messages }))

    const { getByTestId, queryByTestId } = render(<LastMessage conversationId={conversation.id} />)

    expect(queryByTestId('loading')).not.toBeInTheDocument()
    expect(getByTestId('body')).toHaveTextContent('Bar')
    expect(getByTestId('date')).toBeInTheDocument()
  })

  it('should return null if no message', async () => {
    mockedUseMessages.mockImplementation(() => ({ isLoading: false, data: [] }))

    const { queryByTestId } = render(<LastMessage conversationId={conversation.id} />)

    expect(queryByTestId('loading')).not.toBeInTheDocument()
    expect(queryByTestId('body')).not.toBeInTheDocument()
    expect(queryByTestId('date')).not.toBeInTheDocument()
  })
})
