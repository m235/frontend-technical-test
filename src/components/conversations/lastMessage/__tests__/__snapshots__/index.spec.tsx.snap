// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components:conversations:lastMessage should render default correctly 1`] = `
<div>
  <span
    class="sc-bczRLJ eockqr"
    data-testid="loading"
  />
</div>
`;
